crates/pevm/src/chain/optimism.rs

```diff
@@ -189,45 +187,34 @@ impl PevmChain for PevmOptimism {
-                            // TODO: Return proper errors instead of panic-ing.
-                            let account_maybe =
-                                tx_result.state.get(&tx.from).expect("Sender not found");
-                            let account = account_maybe.as_ref().expect("Sender not changed");
-                            let receipt = OpDepositReceipt {
-                                inner: receipt,
-                                deposit_nonce: (spec_id >= SpecId::CANYON)
-                                    .then_some(account.nonce - 1),
-                                deposit_receipt_version: (spec_id >= SpecId::CANYON).then_some(1),
-                            };
-                            OpReceiptEnvelope::Deposit(receipt.with_bloom())
-                        }
-                    })
-                })
-                .collect::<Result<_, _>>()?;
-
-        // 2. Create a trie then calculate the root hash
-        // We use [BTreeMap] because the keys must be sorted in ascending order.
-        let trie_entries: BTreeMap<_, _> = receipt_envelopes
-            .iter()
-            .enumerate()
-            .map(|(index, receipt)| {
-                let key_buffer = alloy_rlp::encode_fixed_size(&index);
-                let mut value_buffer = Vec::new();
-                receipt.encode_2718(&mut value_buffer);
-                (key_buffer, value_buffer)
+        let mut trie_entries = txs
+            .txns()
+            .zip(tx_results.iter())
+            .map(|(tx, tx_result)| {
+                let receipt = tx_result.receipt.clone();
+                match tx.inner.inner.tx_type() {
+                    OpTxType::Legacy => OpReceiptEnvelope::Legacy(receipt.with_bloom()),
+                    OpTxType::Eip2930 => OpReceiptEnvelope::Eip2930(receipt.with_bloom()),
+                    OpTxType::Eip1559 => OpReceiptEnvelope::Eip1559(receipt.with_bloom()),
+                    OpTxType::Eip7702 => OpReceiptEnvelope::Eip7702(receipt.with_bloom()),
+                    OpTxType::Deposit => {
+                        // TODO: Return proper errors instead of panic-ing.
+                        let account_maybe =
+                            tx_result.state.get(&tx.from).expect("Sender not found");
+                        let account = account_maybe.as_ref().expect("Sender not changed");
+                        let receipt = OpDepositReceipt {
+                            inner: receipt,
+                            deposit_nonce: (spec_id >= SpecId::CANYON).then_some(account.nonce - 1),
+                            deposit_receipt_version: (spec_id >= SpecId::CANYON).then_some(1),
+                        };
+                        OpReceiptEnvelope::Deposit(receipt.with_bloom())
+                    }
+                }
             })
-            .collect();
+            .enumerate()
+            .map(|(index, receipt)| (alloy_rlp::encode_fixed_size(&index), receipt.encoded_2718()))
```

Comment:
Just curious, how does `.encoded_2718` accept zero arguments? 

I thought it always accept one argument? `receipt.encode_2718(&mut value_buffer);`

Reply:
Yes! You right, but this method is `.encoded_2718` :)

It is basically:
```rust
fn encoded_2718(&self) -> Vec<u8> {
    let mut buf = Vec::with_capacity(self.encode_2718_len());
    self.encode_2718(&mut buf);
    buf
}
```

Reply:
oh, cool. I didn't notice the difference between `encoded_2718` and  `encode_2718` (the letter `d`)

Reply:
Also, it should be faster (but it's not worth it here), as we will allocate precisely once (`Vec::with_capacity`) and it won't reallocate